_posts/2018-01-14-smaller-docker-images.md

```diff
@@ -0,0 +1,405 @@
+  "dependencies": {
+    "express": "^4.16.2"
+  },
+  "scripts": {
+    "start": "node index.js"
+  }
+}
+```
+
+You can package this application with the following `Dockerfile`:
+
+```dockerfile
+FROM node:8
+
+EXPOSE 3000
+WORKDIR /app
+COPY package.json .
+RUN yarn install
+COPY index.js .
+
+CMD ["yarn", "start"]
+```
+
+You can build the image with:
+
+```bash
+$ docker build -t node-multi-stage .
+```
+
+And you can test that it works correctly with:
+
+```bash
+$ docker run -p 3000:3000 -d node-multi-stage
+```
+
+You should be able to visit [http://localhost:3000](http://localhost:3000) and be greeted by _"Hello World!"_.
+
+There're two `COPY` and one `RUN` statements in the `Dockerfile`. So you should expect to see at least 3 layers more than the base image:
+
+```bash
+$ docker history node-multi-stage
+IMAGE               CREATED             CREATED BY                                      SIZE
+813b579d5aa5        9 seconds ago       /bin/sh -c #(nop)  CMD ["yarn" "start"]         0B
+8bf622b3e30b        9 seconds ago       /bin/sh -c #(nop) COPY file:c370ea0f74a3d7d7…   193B
+79f29e6f7704        9 seconds ago       /bin/sh -c yarn install                         4.11MB
+180851af09a8        13 seconds ago      /bin/sh -c #(nop) COPY file:294c96bce3a399fe…   171B
+500a9fbef90e        13 seconds ago      /bin/sh -c #(nop) WORKDIR /app                  0B
+78b28027dfbf        13 seconds ago      /bin/sh -c #(nop)  EXPOSE 3000                  0B
+b87c2ad8344d        10 days ago         /bin/sh -c #(nop)  CMD ["node"]                 0B
+<missing>           10 days ago         /bin/sh -c set -ex   && for key in     6A010…   4.17MB
+<missing>           10 days ago         /bin/sh -c #(nop)  ENV YARN_VERSION=1.3.2       0B
+<missing>           10 days ago         /bin/sh -c ARCH= && dpkgArch="$(dpkg --print…   56.9MB
+<missing>           10 days ago         /bin/sh -c #(nop)  ENV NODE_VERSION=8.9.4       0B
+<missing>           4 weeks ago         /bin/sh -c set -ex   && for key in     94AE3…   129kB
+<missing>           4 weeks ago         /bin/sh -c groupadd --gid 1000 node   && use…   335kB
+<missing>           4 weeks ago         /bin/sh -c set -ex;  apt-get update;  apt-ge…   324MB
+<missing>           4 weeks ago         /bin/sh -c apt-get update && apt-get install…   123MB
+<missing>           4 weeks ago         /bin/sh -c set -ex;  if ! command -v gpg > /…   0B
+<missing>           4 weeks ago         /bin/sh -c apt-get update && apt-get install…   44.6MB
+<missing>           4 weeks ago         /bin/sh -c #(nop)  CMD ["bash"]                 0B
+<missing>           4 weeks ago         /bin/sh -c #(nop) ADD file:1dd78a123212328bd…   123MB
+```
+
+Instead the resulting image has six new layers: one for each statement in your `Dockerfile`.
+
+Let's try the multi stage Docker build.
+
+You will use the same `Dockerfile` above, but twice:
+
+```dockerfile
+FROM node:8 as build
+
+WORKDIR /app
+COPY package.json .
+RUN yarn install
```

Comment:
why not also run the tests, and then prune the dev dependencies?
see https://github.com/UKHomeOffice/dq-raw-message-retrieval/blob/master/app/Dockerfile for an example
but its basically adding 
```
RUN npm test
RUN npm prune --production
```
(or yarn if you prefer)

Reply:
The intent is to keep the conversation focussed on making the image smaller. While `prune` will effectively make the image smaller, it's an optimisation that happens independently of Docker.

I'd rather leave fewer Docker statements and focus on the process of making smaller images if that makes sense.

Reply:
it does, the interesting thing is that it would demonstrate is that you can have dev-dependencies, or otherwise more files, remove them in a following step which would normally be frowned upon can be acomplished through this pattern